### ISSUE TITLE
`unnecessary_lazy_evaluations` and `once_cell::sync::Lazy`

### ISSUE THREAD

**Issue by arifd** [rust-lang/rust-clippy#10462]
### Summary

I believe this lint should have a specialization for the common case when a `Lazy` is being de-referenced. e.g. unwrapping lazily constructed defaults

### Lint Name

unnecessary_lazy_evaluations

### Reproducer

I tried this code:

```rust
use once_cell::sync::Lazy;

static FOO: Lazy<i32> = Lazy::new(Default::default);

fn main() {
    let foo = Some(1);
    let foo = foo.as_ref().unwrap_or_else(|| &FOO);
    println!("{foo}");
}
```

I saw this happen:

```
warning: unnecessary closure used to substitute value for `Option::None`
 --> src/main.rs:7:15
  |
7 |     let foo = foo.as_ref().unwrap_or_else(|| &FOO);
  |               ^^^^^^^^^^^^^-----------------------
  |                            |
  |                            help: use `unwrap_or(..)` instead: `unwrap_or(&FOO)`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_lazy_evaluations
  = note: `#[warn(clippy::unnecessary_lazy_evaluations)]` on by default

```

I expected to see this happen:

No problimo

### Version

```text
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=58c7612aed45b666f3cd499ebe3a60ce
```
